### ISSUE TITLE
Gateway should not attempt to parse query string directly

### ISSUE THREAD

**Issue by yellow1912** [thephpleague/omnipay#358]
I think the gateway should attempt to use parse_str or access the request query directly like this:

$this->httpRequest->query->get('order')

I think there are cases where different gateways expect different format of $data passed into the Response, I think this should not be case. Data should be uniform, it's the developers' job to make sure the data passed in is uniform (or we could use Adapter)

This makes it very very difficult to test as well as manipulating the flow of the payment. For example, I may want to authorize a payment but do not completeAuthorize it before the order is confirmed, so I may catch the redirect data and use it later.

**Comment by delatbabel** [comment]
Might be a good change for v3.0

**Comment by barryvdh** [comment]
Using the query string makes it possible to complete purchases from a return/notify url in a uniform way. Otherwise the developer would have to parse the query string differently for each gateway. So imho it should use the query string directly.

**Comment by barryvdh** [comment]
Well, not directly obviously, but the Symfony request object is injected to the gateway, probably for this reason. In v3 we'll use the PSR-7 ServerRequestInterface for this.

**Comment by yellow1912** [comment]
Yup. While I completely understand why it is done that way, the Response handler should not parse the query string itself, it should expect some kind of object, array data passed to it. The parsing should happen elsewhere at least. That will give us the ability to mock data or to pre-process data. I like the idea with ServerRequestInterface.

**Comment by barryvdh** [comment]
You can already mock the data, just pass in your own Request object here: 
https://github.com/thephpleague/omnipay-common/blob/v2.5.0/src/Omnipay/Common/AbstractGateway.php#L68-L73 or here: https://github.com/thephpleague/omnipay-common/blob/v2.5.0/src/Omnipay/Common/GatewayFactory.php#L100-L109

**Comment by judgej** [comment]
Should I mention `Psr\Http\Message\ServerRequestInterface` (PSR-7, OmniPay v3) or is that a step too far?
